client/main/dataset.js

```diff
@@ -895,6 +938,188 @@ const DataSetModel = Backbone.Model.extend({
+                transformPB.description = transform.description;
+            else
+                transformPB.description = '';
+
+            if ('formula' in values)
+                transformPB.formula = values.formula;
+            else if ( ! newTransform)
+                transformPB.formula = transform.formula;
+            else
+                transformPB.formula = [ '' ];
+
+            datasetPB.schema.transforms.push(transformPB);
+        }
+
+        let request = new coms.Messages.ComsMessage();
+        request.payload = datasetPB.toArrayBuffer();
+        request.payloadType = 'DataSetRR';
+        request.instanceId = this.attributes.instanceId;
+
+        return coms.send(request).then(response => {
+            let datasetPB = coms.Messages.DataSetRR.decode(response.payload);
+            if (datasetPB.incSchema) {
+
+                let changed = Array(datasetPB.schema.transforms.length);
+                let changes = Array(datasetPB.schema.transforms.length);
+                let nCreated = 0;
+
+                for (let i = 0; i < datasetPB.schema.transforms.length; i++) {
+                    let transformPB = datasetPB.schema.transforms[i];
+                    let id = transformPB.id;
+                    let transform = this.getTransformById(id);
+
+                    let newName = transformPB.name;
+
+                    let created;
+                    let oldName;
+                    let oldMessage;
+
+                    if (transform !== undefined) {
+                        created = false;
+                        oldName = transform.name;
+                        oldMessage = transform.formulaMessage;
+                        this._readTransformPB(transform, transformPB);
+                    }
+                    else {
+                        created = true;
+                        oldName = transformPB.name;
+                        oldMessage = '';
+                        transform = { };
+                        nCreated++;
+                        this._readTransformPB(transform, transformPB);
+                        this.attributes.transforms.push(transform);
+                    }
+                    let nameChanged = (oldName !== transformPB.name);
+
+                    changed[i] = transformPB.name;
+                    changes[i] = {
+                        id: id,
+                        name: transform.name,
+                        oldName: oldName,
+                        nameChanged: nameChanged,
+                        created: created,
+                        formulaMessageChanged: transform.formulaMessage !== oldMessage
+                    };
+                }
+
+                for (let change of changes) {
+                    if (change.created)
+                        this.trigger('transformAdded', { id: change.id });
+                }
+
+                this.trigger('transformsChanged', { changed, changes });
+            }
+        }).catch((error) => {
```

Comment:
use `.done()`:

https://github.com/kriskowal/q#the-end

Reply:
Is not necessary because these promises return.